docs/design-docs/08-package-variant.md

```diff
@@ -0,0 +1,533 @@
+over-parameterization. The preferred approach is configuration injection, as
+described below, since it allows inputs to adhere to a well-defined, reusable
+schema, rather than simple key/value pairs.
+
+### KRM Function Pipeline[^notimplemented]
+In the manual workflow, one of the ways we edit packages is by running KRM
+functions imperatively. `PackageVariant` offers a similar capability, by
+allowing the user to list a KRM function pipeline similar to the `Kptfile`
+mutators pipeline. By default, these KRM functions are not added to the
+`Kptfile` pipeline, but there is an option to do so.
+
+Creating a namespace and/or setting the namespace for a particular package is a
+very common operation. However, since namespace provisioning in a cluster is a
+privileged operation, the deployer of a package may not have the authority to
+provision a namespace. For this reason, upstream packages should not directly
+include `Namespace` resources if they want to be truly reusable.
+
+The KRM function pipeline provides an easy mechanism for the deployer to set the
+namespace, or even to create one if their downstream package application
+pipeline allows it.[^setns]
+
+### Configuration Injection[^pdc]
+Adding values to the package context, or executing functions with their
+configuration listed in the `PackageVariant`, works for values that are under
+control of the author of the `PackageVariant` resource. However, in more
+advanced use cases, we may need to specialize the package based upon other
+contextual information. This particularly comes into play when the user
+deploying the workload does not have direct control over the context in which
+it is being deployed. For example, one part of the organization may manage the
+infrastructure - that is, the cluster in which we are deploying the workload -
+and another part the actual workload. We would like to be able to pull in inputs
+specified by the infrastructure team automatically, based the cluster to which
+we are deploying the workload, or perhaps the region in which that cluster is
+deployed.
+
+To facilitate this, the package variant controller can "inject" configuration
+directly into the package. This means it will use information specific to this
+instance of the package to lookup a resource in the Porch cluster, and copy that
+information into the package. Of course, the package has to be ready to receive
+this information. So, there is a protocol for facilitating this dance:
+- Packages may contain resources annotated with `kpt.dev/config-injection`
+- Usually, these will also be `config.kubernetes.io/local-config` resources, as
+  they are likely just used by local functions as input. But this is not
+  mandatory.
+- The package variant controller will look for any resource in the Kubernetes
+  cluster matching the Group, Version, and Kind of the package resource, and
+  satisfying the *injection selector*.
+- The package variant controller will copy the `spec` field from the matching
```

Comment:
I am not 100% convinced that copying *only* `spec` is the right thing to do. We for sure do not want to copy `metadata` (we don't want to overwrite in-package `metadata`). But everything else is game. For example, if we want to inject a resource representing a cluster that is *discovered*, might that information be in the `status`? Or, if we define a resource that is not used to actuate anything, but really just contains information, does that really belong in a `spec` field?

Reply:
On the other hand I don't want to end up with a bunch of status cruft. Hmm. Maybe we see how this goes for a while, and later if we want to say, add a list of fields to copy, we can do that.

Reply:
I agree that spec seems to be the only data that should be copied, assuming that resources follow the convention of using only `metadata`, `spec`, and `status`.